e2e/tenv.ts

```diff
@@ -0,0 +1,126 @@
+/* eslint-disable import/no-nodejs-modules */
+import childProcess from 'child_process';
+import fs from 'fs/promises';
+import { createServer } from 'http';
+import { AddressInfo } from 'net';
+import path from 'path';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { Repeater } from '@repeaterjs/repeater';
+
+let leftovers: Proc[] = [];
+afterAll(async () => {
+  await Promise.allSettled(
+    leftovers.map(proc => {
+      proc.kill();
+      return proc.waitForExit;
+    }),
+  );
+  leftovers = [];
+});
+
+export interface Proc {
+  stdout: Repeater<string>;
+  stderr: Repeater<string>;
+  kill(code?: number): void;
+  waitForExit: Promise<void>;
+}
+
+export interface Tenv {
+  getAvailablePort(): number;
+  spawn(cmd: string, ...args: (string | number)[]): Promise<Proc>;
+  fileExists(filePath: string): Promise<boolean>;
+  readFile(filePath: string): Promise<string>;
+  // If the file doesn't exist, is a noop.
+  deleteFile(filePath: string): Promise<void>;
+}
+
+export function createTenv(cwd: string): Tenv {
+  return {
+    getAvailablePort() {
+      const server = createServer();
+      server.listen(0);
+      const { port } = server.address() as AddressInfo;
+      server.close();
+      return port;
+    },
```

Comment:
Couldn't be better to give the `0` port to mesh-serve and have a util function that parses the port from the stdout ?

Reply:
Mesh doesn't support port 0, even if it did - parsing stdout is tedious. This is more deterministic.

Reply:
Ok, I just worried that the port is getting grabbed by an other process (a parrallel test run for example) between this function being run and the actual use of the port.

Reply:
It shouldn't, JS is not concurrent. It could happen if running multiple jest instances. But that I wouldnt worry about for now.